nomt/src/beatree/allocator/free_list.rs

```diff
@@ -369,15 +348,55 @@ fn decode_free_list_page(page: FatPage) -> (PageNumber, Vec<PageNumber>) {
 fn encode_free_list_page(page_pool: &PagePool, prev: PageNumber, pns: &[PageNumber]) -> FatPage {
     let mut page = page_pool.alloc_fat_page();
 
-    page[0..4].copy_from_slice(&prev.0.to_le_bytes());
-    page[4..6].copy_from_slice(&(pns.len() as u16).to_le_bytes());
+    {
+        let mut e = FreeListPageMut(&mut page[..]);
+        e.set_prev_pn(prev);
+        e.set_item_count(pns.len() as u16);
+        for (i, pn) in pns.into_iter().enumerate() {
+            e.set_item(i, *pn);
+        }
+    }
+
+    page
+}
+
+// A free page is laid out in the following form:
+// + prev free page : u32
+// + item_count : u16
+// + free pages : [u32; item_count]
+struct FreeListPageRef<'a>(&'a [u8]);
+struct FreeListPageMut<'a>(&'a mut [u8]);
+
+impl<'a> FreeListPageRef<'a> {
+    fn prev_pn(&self) -> PageNumber {
+        PageNumber(u32::from_le_bytes(self.0[0..4].try_into().unwrap()))
+    }
+
+    fn item_count(&self) -> u16 {
+        u16::from_le_bytes(self.0[4..6].try_into().unwrap())
+    }
 
-    for (i, pn) in pns.into_iter().enumerate() {
+    fn item(&self, i: usize) -> PageNumber {
         let start = 6 + i * 4;
-        page[start..start + 4].copy_from_slice(&pn.0.to_le_bytes());
+        PageNumber(u32::from_le_bytes(
+            self.0[start..start + 4].try_into().unwrap(),
+        ))
     }
+}
 
-    page
+impl<'a> FreeListPageMut<'a> {
+    fn set_prev_pn(&mut self, pn: PageNumber) {
+        self.0[0..4].copy_from_slice(&pn.0.to_le_bytes());
+    }
+
+    fn set_item_count(&mut self, count: u16) {
+        self.0[4..6].copy_from_slice(&count.to_le_bytes());
+    }
+
+    fn set_item(&mut self, i: usize, pn: PageNumber) {
```

Comment:
This will throw if you try to access index outside of the underlying array - is that OK?

Reply:
should be fine. Less for for a public API, but this is not meant to be one. The whole change is just trying to improve legibility of the code.

Reply:
Yeah. In a public API it's worth at least documenting, but since it's private it should be fine.